### PR TITLE
Allow blocking some email domains from registering an account

### LINKED ISSUE #13628: [Feature] block-/allow-list for emails
It would be nice to reject email addresses from ... gmail, ru-tld domain, ...

I noticed spam accounts created with specific domains ....

ref: https://github.com/go-gitea/gitea/blob/f915161a2f1f4a1c436e7a96cfc88f61ae8fbce6/models/user_mail.go#L46

similar: #13610

### PR DESCRIPTION
**Pull request by gouttegd**
Gitea allows to whitelist email domains so that only email addresses from certain domains are allowed to register an account, but does not currently allows to do the opposite: blacklisting email domains so that addresses from certain domains are *forbidden* to register an account.

The idea has been briefly mentioned in the discussion about issue #6350, but never implemented. This PR does that.

The rationale is that, in my experience of running a Gitea instance, *a single email domain* is responsible for *most* of the spam accounts, and for *all* of the spam accounts that manage to get past the email confirmation step. So on top of the other spam mitigation measures already available (email confirmation, CAPTCHA, etc.), having the option to block a particularly annoying domain would be helpful.

close #13628

### CODE REVIEW

File: docs/content/doc/help/faq.en-us.md [hunk 1]
```diff
@@ -120,13 +120,14 @@ For more information, refer to Gitea's [API docs]({{< relref "doc/developers/api
...
+1. By whitelisting or blacklisting certain email domains
+2. By only whitelisting certain domains with OpenID (see below)
+3. Setting `ENABLE_CAPTCHA` to `true` in your `app.ini` and properly configuring `RECAPTCHA_SECRET` and `RECAPTCHA_SITEKEY`
+4. Settings `DISABLE_REGISTRATION` to `true` and creating new users via the [CLI]({{< relref "doc/usage/command-line.en-us.md" >}}), [API]({{< relref "doc/developers/api-usage.en-us.md" >}}), or Gitea's Admin UI
 
-### Only allow certain email domains
+### Only allow/block certain email domains
 
-You can configure `EMAIL_DOMAIN_WHITELIST` in your app.ini under `[service]`
+You can configure `EMAIL_DOMAIN_WHITELIST` and/or `EMAIL_DOMAIN_BLACKLIST` in your app.ini under `[service]`
```
Comment (6543) [inline docs/content/doc/help/faq.en-us.md]:
```suggestion
You can configure `EMAIL_DOMAIN_WHITELIST` or `EMAIL_DOMAIN_BLACKLIST` in your app.ini under `[service]`
```
using a blocklist if you already use a whitelist is nonecence to mee

File: custom/conf/app.example.ini [hunk 2]
```diff
@@ -617,6 +617,8 @@ REGISTER_MANUAL_CONFIRM = false
 ; List of domain names that are allowed to be used to register on a Gitea instance
 ; gitea.io,example.com
 EMAIL_DOMAIN_WHITELIST =
+; Comma-separated list of domain names that are not allowed to be used to register on a Gitea instance
+EMAIL_DOMAIN_BLACKLIST =
```
Comment (6543) [inline custom/conf/app.example.ini]: there where some complains about blacklist vs blocklist ... lets just stick with **blocklist** ...

File: modules/forms/user_form.go [hunk 3]
```diff
@@ -120,6 +120,30 @@ func (f RegisterForm) IsEmailDomainWhitelisted() bool {
 	return false
 }
 
+// IsEmailDomainBlacklisted validates that the email address
+// does not come from a domain that has been blacklisted.
+// In the absence of a blacklist, all addresses are accepted.
+func (f RegisterForm) IsEmailDomainBlacklisted() bool {
+	if len(setting.Service.EmailDomainBlacklist) == 0 {
+		return false
```
Comment (6543) [inline modules/forms/user_form.go]: looks like code de-dublication can be done

File: custom/conf/app.example.ini (hunk 2 above)
Comment (gouttegd) [reply custom/conf/app.example.ini]: Fine with me. I have used *blacklist* solely for consistency with the `BLACKLIST_URIS` setting for OpenID.

Comment (6543) [reply custom/conf/app.example.ini]: yes keep the old one since it would break things but if we can avoid blacklist ...

**Review by 6543 — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**